### PR TITLE
added dot timer features for brd, ast, whm, sge, and sch

### PR DESCRIPTION
**Pull request by lhn1703**
Extremely naughty features: refreshes dot for you. 

Don't wanna mess around with Eukrasian Dosis clusterfuck, maybe someone else can implement that :)

### CODE REVIEW

**Comment by MKhayle** [comment]
I need to think about those ones
or rather, wait until we get to the attributes rework?

**Comment by vitharr137** [comment]
For added naughtiness, the details for BaseCooldown, CooldownRemaining, and CastTime can be combined to make a dynamic buffer time that checks ahead to see if you should refresh rather than looking at a static integer for it, like so:
```
double remainingGCD = GetCooldown(AST.Malefic).CooldownRemaining;
double castTime = GetCooldown(AST.Malefic).CastTime;
double maintenance = remainingGCD + castTime;
```
`(!TargetHasEffect(AST.Debuffs.Combust3) || FindTargetEffect(AST.Debuffs.Combust3)?.RemainingTime <= maintenance)`

which also respects Lightspeed and such

For Bard, I needed to add in a static amount to account for Iron Jaw's rather long animation lock before it actually triggered the refresh:
```
double gcd = GetCooldown(BRD.HeavyShot).BaseCooldown;
double remainingGCD = GetCooldown(BRD.HeavyShot).CooldownRemaining;
double bufferGCD = 0.71f;
double totalGCD = remainingGCD + gcd;
double maintGCD = remainingGCD + gcd + bufferGCD;
```

**Comment by lhn1703** [comment]
> Extremely naughty features: refreshes dot for you.
> 
> Don't wanna mess around with Eukrasian Dosis clusterfuck, maybe someone else can implement that :)

Turns out just replacing dosis and checking for eukrasia buff is enough to implement dot feature.

**Comment by MKhayle** [comment]
It's not going to get "harder" but I feel that's exactly some of the features we'll have to decide whether it's expanded, secret or accessibility

File: XIVComboExpanded/CustomComboPreset.cs [hunk 1]
```diff
@@ -142,6 +142,10 @@ public enum CustomComboPreset
     [CustomComboInfo("Benefic II to Benefic Level Sync", "Replace Benefic 2 with Benefic when below level 26 in synced content.", AST.JobID)]
     AstrologianBeneficSyncFeature = 3326,
 
+    [SecretCustomCombo]
+    [CustomComboInfo("AST DoT Feature", "Replace Malefic with Combust when it is about to run out.", AST.JobID)]
```
Comment (kaedys) [inline XIVComboExpanded/CustomComboPreset.cs:146]:
```suggestion
    [CustomComboInfo("Auto Combust", "Replace Malefic with Combust when it is about to run out.", AST.JobID)]
```

File: XIVComboExpanded/CustomComboPreset.cs [hunk 2]
```diff
@@ -1084,6 +1092,10 @@ public enum CustomComboPreset
     [CustomComboInfo("Toxikon into Phlegma Feature", "Replace Toxikon with Phlegma when charges are available.", SGE.JobID)]
     SageToxikonPhlegma = 4011,
 
+    [SecretCustomCombo]
+    [CustomComboInfo("SGE DoT Feature", "Replace Dosis with Eukrasia when Eukrasian Dosis is about to run out.", SGE.JobID)]
```
Comment (kaedys) [inline XIVComboExpanded/CustomComboPreset.cs:1096]:
```suggestion
    [CustomComboInfo("Auto Eukrasian Dosis", "Replace Dosis with Eukrasia when Eukrasian Dosis is about to run out.", SGE.JobID)]
```

File: XIVComboExpanded/CustomComboPreset.cs [hunk 3]
```diff
@@ -1195,6 +1207,10 @@ public enum CustomComboPreset
     [CustomComboInfo("Adloquium Level Sync", "Replace Adloquium with Physick when below level 30 in synced content.", SCH.JobID)]
     ScholarAdloquiumSyncFeature = 2810,
 
+    [SecretCustomCombo]
+    [CustomComboInfo("SCH DoT Feature", "Replace Malefic with Combust when it is about to run out.", SCH.JobID)]
```
Comment (kaedys) [inline XIVComboExpanded/CustomComboPreset.cs:1211]:
```suggestion
    [CustomComboInfo("Auto Biolysis", "Replace Broil with Biolysis when it is about to run out.", SCH.JobID)]
```

File: XIVComboExpanded/CustomComboPreset.cs [hunk 4]
```diff
@@ -1426,6 +1442,10 @@ public enum CustomComboPreset
     [CustomComboInfo("Glare4 Feature", "Replace Glare 3 with Glare 4 when a stack is available.", WHM.JobID)]
     WhiteMageGlare4Feature = 2407,
 
+    [SecretCustomCombo]
+    [CustomComboInfo("WHM DoT Feature", "Replace Stone or Glare 3 with Aero or Dia when it is about to run out.", WHM.JobID)]
```
Comment (kaedys) [inline XIVComboExpanded/CustomComboPreset.cs:1446]:
```suggestion
    [CustomComboInfo("Auto Dia", "Replace Glare with Dia when it is about to run out.", WHM.JobID)]
```

**Review by kaedys — COMMENTED** [review]
The Scholar needs updated regardless, but I also recommended some rewording of the names and descriptions.  I'm also definitely stealing this for my local branch, because gods I'm bad at refreshing my DoTs sometimes.

File: XIVComboExpanded/Combos/AST.cs [hunk 5]
```diff
@@ -86,6 +89,17 @@ protected override uint Invoke(uint actionID, uint lastComboMove, float comboTim
...
             var gauge = GetJobGauge<ASTGauge>();
 
+            if (IsEnabled(CustomComboPreset.AstrologianDoTFeature))
+            {
+                var combust = FindTargetEffect(AST.Debuffs.Combust);
+                var combust2 = FindTargetEffect(AST.Debuffs.Combust2);
+                var combust3 = FindTargetEffect(AST.Debuffs.Combust3);
+
+                // have to explicitly check all variants of the dot for some reason else spaghetti code ensues
+                if (combust?.RemainingTime < 2.8 || combust2?.RemainingTime < 2.8 || combust3?.RemainingTime < 2.8)
```
Comment (kaedys) [inline XIVComboExpanded/Combos/AST.cs:99]:
```suggestion
                if (!(combust?.RemainingTime > 2.8 || combust2?.RemainingTime > 2.8 || combust3?.RemainingTime > 2.8))
```

File: XIVComboExpanded/Combos/SCH.cs [hunk 6]
```diff
@@ -243,3 +251,27 @@ protected override uint Invoke(uint actionID, uint lastComboMove, float comboTim
...
+        if (actionID == SCH.Ruin || actionID == SCH.Broil || actionID == SCH.Broil2 || actionID == SCH.Broil3 || actionID == SCH.Broil4)
+        {
+            if (IsEnabled(CustomComboPreset.ScholarDoTFeature))
+            {
+                var bio = FindTargetEffect(SCH.Debuffs.Bio);
+                var bio2 = FindTargetEffect(SCH.Debuffs.Bio2);
+                var biolysis = FindTargetEffect(SCH.Debuffs.Biolysis);
+
+                // have to explicitly check all variants of the dot for some reason else spaghetti code ensues
+                if (bio?.RemainingTime < 2.8 || bio2?.RemainingTime < 2.8 || biolysis?.RemainingTime < 2.8)
```
Comment (kaedys) [inline XIVComboExpanded/Combos/SCH.cs:270]:
```suggestion
                if (!(bio?.RemainingTime > 2.8 || bio2?.RemainingTime > 2.8 || biolysis?.RemainingTime > 2.8))
```

File: XIVComboExpanded/Combos/SGE.cs [hunk 7]
```diff
@@ -80,8 +86,22 @@ internal class SageDosis : CustomCombo
...
+            {
+                var eurkasiandosis = FindTargetEffect(SGE.Debuffs.EukrasianDosis);
+                var eurkasiandosis2 = FindTargetEffect(SGE.Debuffs.EukrasianDosis2);
+                var eurkasiandosis3 = FindTargetEffect(SGE.Debuffs.EukrasianDosis3);
+
+                if (HasEffect(SGE.Buffs.Eukrasia))
+                    return OriginalHook(SGE.Dosis);
+
+                // have to explicitly check all variants of the dot for some reason else spaghetti code ensues
+                if (eurkasiandosis?.RemainingTime < 2.8 || eurkasiandosis2?.RemainingTime < 2.8 || eurkasiandosis3?.RemainingTime < 2.8)
```
Comment (kaedys) [inline XIVComboExpanded/Combos/SGE.cs:101]:
```suggestion
                if (!(eurkasiandosis?.RemainingTime > 2.8 || eurkasiandosis2?.RemainingTime > 2.8 || eurkasiandosis3?.RemainingTime > 2.8))
```

File: XIVComboExpanded/Combos/WHM.cs [hunk 8]
```diff
@@ -200,6 +203,17 @@ protected override uint Invoke(uint actionID, uint lastComboMove, float comboTim
...
         if (actionID == WHM.Stone)
         {
+            if (IsEnabled(CustomComboPreset.WhiteMageDoTFeature))
+            {
+                var aero = FindTargetEffect(WHM.Debuffs.Aero);
+                var aero2 = FindTargetEffect(WHM.Debuffs.Aero2);
+                var dia = FindTargetEffect(WHM.Debuffs.Dia);
+
+                // have to explicitly check all variants of the dot for some reason else spaghetti code ensues
+                if (aero?.RemainingTime < 2.8 || aero2?.RemainingTime < 2.8 || dia?.RemainingTime < 2.8)
```
Comment (kaedys) [inline XIVComboExpanded/Combos/WHM.cs:213]:
```suggestion
                if (!(aero?.RemainingTime > 2.8 || aero2?.RemainingTime > 2.8 || dia?.RemainingTime > 2.8))
```

File: XIVComboExpanded/Combos/WHM.cs [hunk 9]
```diff
@@ -200,6 +203,17 @@ protected override uint Invoke(uint actionID, uint lastComboMove, float comboTim
     {
         if (actionID == WHM.Stone)
         {
+            if (IsEnabled(CustomComboPreset.WhiteMageDoTFeature))
```
Comment (kaedys) [inline XIVComboExpanded/Combos/WHM.cs:206]:
```suggestion
            if (IsEnabled(CustomComboPreset.WhiteMageDoTFeature) && TargetIsEnemy())
```

File: XIVComboExpanded/Combos/SGE.cs [hunk 10]
```diff
@@ -80,8 +86,22 @@ internal class SageDosis : CustomCombo
 
     protected override uint Invoke(uint actionID, uint lastComboMove, float comboTime, byte level)
     {
-        if (actionID == SGE.Dosis)
+        if (actionID == SGE.Dosis || actionID == SGE.Dosis2 || actionID == SGE.Dosis3)
         {
+            if (IsEnabled(CustomComboPreset.SageDoTFeature))
```
Comment (kaedys) [inline XIVComboExpanded/Combos/SGE.cs:91]:
```suggestion
            if (IsEnabled(CustomComboPreset.SageDoTFeature) && TargetIsEnemy())
```

File: XIVComboExpanded/Combos/SCH.cs [hunk 11]
```diff
@@ -243,3 +251,27 @@ protected override uint Invoke(uint actionID, uint lastComboMove, float comboTim
...
+
+internal class ScholarRuin : CustomCombo
+{
+    protected internal override CustomComboPreset Preset { get; } = CustomComboPreset.SchAny;
+
+    protected override uint Invoke(uint actionID, uint lastComboMove, float comboTime, byte level)
+    {
+        if (actionID == SCH.Ruin || actionID == SCH.Broil || actionID == SCH.Broil2 || actionID == SCH.Broil3 || actionID == SCH.Broil4)
+        {
+            if (IsEnabled(CustomComboPreset.ScholarDoTFeature))
```
Comment (kaedys) [inline XIVComboExpanded/Combos/SCH.cs:263]:
```suggestion
            if (IsEnabled(CustomComboPreset.ScholarDoTFeature) && TargetIsEnemy())
```

File: XIVComboExpanded/Combos/AST.cs [hunk 12]
```diff
@@ -86,6 +89,17 @@ protected override uint Invoke(uint actionID, uint lastComboMove, float comboTim
         {
             var gauge = GetJobGauge<ASTGauge>();
 
+            if (IsEnabled(CustomComboPreset.AstrologianDoTFeature))
```
Comment (kaedys) [inline XIVComboExpanded/Combos/AST.cs:92]:
```suggestion
            if (IsEnabled(CustomComboPreset.AstrologianDoTFeature) && TargetIsEnemy())
```

**Review by kaedys — COMMENTED** [review]
Few modifications from testing this locally.  I added an enemy check, both to reduce the code running when it shouldn't, and to avoid the button confusingly changing when you don't have a target.

In addition, I swapped the conditional to a negated inversion of the prior condition.  Main reason is that the `?` operator in C# causes all inequality comparisons to return `false` if the root variable is nil, which in effect means that `ability?.RemainingTime < 2.8` returns the same result if it is on the target and high on duration, or if it is entirely missing.  Using `!(ability?.RemainingTime > 2.8)` fixes this, as it means the `false` result from the debuff being entirely missing is negate to a `true`, without affecting the logic if the debuff is present.  Net result is that the feature will now _apply_ the DoT, in addition to simply refreshing it (doesn't apply to Iron Jaws, for obvious reasons, just the healer DoTs).

**Comment by MKhayle** [comment]
are we getting them in accessibility or in expanded
I'd say accessibility tbh

**Comment by MKhayle** [comment]
(btw the `?` thing and workaround is kinda nuts)

**Comment by kaedys** [comment]
Ya, these definitely feel like accessibility.  And yes, `?` is weird.  Still, better than having to explicitly null-check everything first >.<

**Comment by MKhayle** [comment]
I changed the combo & skill names in my local AttributeRework branch by their earliest form (Glare > Stone, Dia > Aero, Boil > Ruin, Biolysis > Bio) so that it remains coherent with AST Malefic & Combust and also is easier to understand for people who don't have high leveled jobs

**Comment by lhn1703** [comment]
I might create a black mage auto thunder feature, but I am conflicted since it is a movement skill and cause you to drop enochian timer.

**Comment by kaedys** [comment]
Ya, that one would be difficult.  It's not sufficient to just check the DoT timer, you need to check remaining Enochian duration as well as the cast time of whatever would be used to refresh it (0 for something like Paradox, very not zero for Despair).

**Comment by MKhayle** [comment]
as a secret combo, yes, making it optimal would be hard
as an accessibility combo, dropping enochian mid-cast is technically worse than dropping DoT uptime but more reasonable and wouldn't feel as bad for the "targeted audience"

**Comment by kaedys** [comment]
TBH, even as accessibility, it feels like a trap rather than an asset.  Dropping element is way WAY worse than lower DoT uptime.  The only way I think it'd make sense is if we also condition it on being in Ice element, since you basically never use enough casts during ice combo to risk dropping element, whereas in Fire, it's _intentional_ that you basically play right up to the end of the timer.

**Pull request merged**